### PR TITLE
fix: Use `.zip` in archive path

### PR DESCRIPTION
**Pull request by dnys1**
Fixes https://github.com/dart-lang/setup-dart/issues/118

Ensures that downloads of the Dart SDK zip file carry the `.zip` extension so that subsequent calls to `Expand-Archive` on Windows succeed.

## Notes

- The diff of `index.mjs` is quite large. I believe this is due to the version of Node I'm using (`20.7.0`). Happy to use a different version to keep the diff small, but wasn't sure which one is currently used. This could be noted in `package.json`  (under [engines](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#engines)), if desired, to ensure alignment with external contributions.

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

**Comment by mit-mit** [comment]
Thanks @dnys1! Would you mind resolving conflicts? Then we'll take a look.

**Comment by dnys1** [comment]
@mit-mit Done 😄

File: lib/main.dart
Comment (devoncarew) [inline lib/main.dart:111]:
```suggestion
      // https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables
```

File: lib/node/actions/tool_cache.dart
Comment (devoncarew) [inline lib/node/actions/tool_cache.dart:11]: nit: can you confirm that this file is dart formatted?

**Review by devoncarew — APPROVED** [review]
Looks good, thanks!

Can you also add a new changelog entry to the changelog file? You can start a new `v1.6.5-wip` entry.

> The diff of index.mjs is quite large. I believe this is due to the version of Node I'm using (20.7.0). Happy to use a different version to keep the diff small, but wasn't sure which one is currently used. This could be noted in package.json (under [engines](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#engines)), if desired, to ensure alignment with external contributions.

fwiw I'm using v20.6.1. If you think that this will reduce diffs then PRs here would be welcome. We're mostly treating the dist/index.mjs file as opaque - we know it works if all the CI tests pass. Do you think specifying a specific version of node would help? I'd be worried about then remembering to rev that version periodically. Plus, many of the diffs in the file may be from the npm packages in use?

**Review by dnys1 — COMMENTED** [review]
Hey @devoncarew, apologies for the delay on this.

> Can you also add a new changelog entry to the changelog file? You can start a new `v1.6.5-wip` entry.

Added 👍

> fwiw I'm using v20.6.1. If you think that this will reduce diffs then PRs here would be welcome. We're mostly treating the dist/index.mjs file as opaque - we know it works if all the CI tests pass. Do you think specifying a specific version of node would help? I'd be worried about then remembering to rev that version periodically. Plus, many of the diffs in the file may be from the npm packages in use?

If you're okay with treating `index.mjs` as opaque, I think that's an equally good option. There could still be benefit in adding an engine reference, which would just be to ensure older (perhaps, deprecated) versions of Node are not used by mistake. In this case, the constraint could be more lax, e.g. `>=20`.

**Comment by ntkme** [comment]
The point of using the same node js version as what's running in CI is to make sure the committed change is consistent with a rebuild on CI - in order to prevent hackers from attempting to inject obfuscated malicious code directly to the JS output.

@dnys1 The CI is probably failing because of this, and you probably have to rebuild locally with the same node version running in CI in order to get it to pass.

**Comment by ntkme** [comment]
I'm blocked by this issue. As I don't have access to directly contribute to this PR, so I opt to take the commits from this PR, and created a new PR with one more additional commit to fix the build issue: https://github.com/dart-lang/setup-dart/pull/134

**Comment by dnys1** [comment]
Thanks @ntkme I hadn't seen the build failure. I've pushed a new commit which runs the formatter, rebuilds, and signs the JS using `npm run all` on Dart 3.4.1.

**Comment by ntkme** [comment]
@dnys1 Thanks for the update. @devoncarew Can you please take another look and see if we can get this merged?

**Comment by kevmoo** [comment]
I'll leave for @devoncarew to hit the merge button!

**Review by devoncarew — APPROVED** [review]

**Comment by devoncarew** [comment]
Thanks for the contribution!

**Pull request merged**
